src/applications/apps.ts

```diff
@@ -383,75 +411,95 @@ function validCustomProps(customProps) {
+  appNameOrConfig: string | RegisterApplicationConfig,
+  appOrLoadApp: Application,
+  activeWhen: Activity,
+  customProps?: ExtraProps | CustomPropsFn<ExtraProps>
+): ApplicationRegistration {
   const usingObjectAPI = typeof appNameOrConfig === "object";
 
-  const registration = {
-    name: null,
-    loadApp: null,
-    activeWhen: null,
-    customProps: null,
+  let unsanitizedRegistration: {
+    name: string;
+    app: Application;
+    activeWhen: Activity;
+    customProps: CustomProps;
   };
 
   if (usingObjectAPI) {
     validateRegisterWithConfig(appNameOrConfig);
-    registration.name = appNameOrConfig.name;
-    registration.loadApp = appNameOrConfig.app;
-    registration.activeWhen = appNameOrConfig.activeWhen;
-    registration.customProps = appNameOrConfig.customProps;
+    unsanitizedRegistration = {
+      name: appNameOrConfig.name,
+      app: appNameOrConfig.app,
+      activeWhen: appNameOrConfig.activeWhen,
+      customProps: appNameOrConfig.customProps,
+    };
   } else {
     validateRegisterWithArguments(
       appNameOrConfig,
       appOrLoadApp,
       activeWhen,
       customProps
     );
-    registration.name = appNameOrConfig;
-    registration.loadApp = appOrLoadApp;
-    registration.activeWhen = activeWhen;
-    registration.customProps = customProps;
+    unsanitizedRegistration = {
+      name: appNameOrConfig,
+      app: appOrLoadApp,
+      activeWhen,
+      customProps,
+    };
   }
 
-  registration.loadApp = sanitizeLoadApp(registration.loadApp);
-  registration.customProps = sanitizeCustomProps(registration.customProps);
-  registration.activeWhen = sanitizeActiveWhen(registration.activeWhen);
-
-  return registration;
+  return {
+    name: unsanitizedRegistration.name,
+    loadApp: sanitizeLoadApp(unsanitizedRegistration.app),
+    customProps: sanitizeCustomProps(unsanitizedRegistration.customProps),
+    activeWhen: sanitizeActiveWhen(unsanitizedRegistration.activeWhen),
+  };
 }
 
-function sanitizeLoadApp(loadApp) {
+function sanitizeLoadApp<ExtraProps extends CustomProps = {}>(
+  loadApp: Application
+): (config: ExtraProps & AppProps) => Promise<LifeCycles<ExtraProps>> {
   if (typeof loadApp !== "function") {
     return () => Promise.resolve(loadApp);
   }
 
   return loadApp;
 }
 
-function sanitizeCustomProps(customProps) {
+function sanitizeCustomProps(customProps?: CustomProps): CustomProps {
   return customProps ? customProps : {};
 }
 
-function sanitizeActiveWhen(activeWhen) {
-  let activeWhenArray = Array.isArray(activeWhen) ? activeWhen : [activeWhen];
-  activeWhenArray = activeWhenArray.map((activeWhenOrPath) =>
-    typeof activeWhenOrPath === "function"
-      ? activeWhenOrPath
-      : pathToActiveWhen(activeWhenOrPath)
+function sanitizeActiveWhen(activeWhen: Activity): ActivityFn {
+  const activeWhenArray: (ActivityFn | string)[] = Array.isArray(activeWhen)
+    ? activeWhen
+    : [activeWhen];
+  const activityFnArray: ActivityFn[] = activeWhenArray.map(
```

Comment:
To facilitate stricter compile-time checks, I created a new runtime variable called `activityFnArray` rather than reassigning `activeWhenArray`.